File: core/meetings/services/meeting-api/src/meeting_api/collector/alloy_stt_telemetry.py
"""ALLOY read adapter for ephemeral per-meeting STT scheduler snapshots."""
from __future__ import annotations

import json
import math
from typing import Any, Iterable


def alloy_stt_telemetry_key(meeting_id: int | str) -> str:
    return f"alloy:stt:telemetry:v1:{meeting_id}"


def _is_non_negative_int(value: Any) -> bool:
    return isinstance(value, int) and not isinstance(value, bool) and value >= 0


def _is_non_negative_number(value: Any, *, optional: bool = False) -> bool:
    if optional and value is None:
        return True
    return (
        isinstance(value, (int, float))
        and not isinstance(value, bool)
        and math.isfinite(float(value))
        and value >= 0
    )


def _is_valid_last_error(value: Any) -> bool:
    if value is None:
        return True
    return (
        isinstance(value, dict)
        and isinstance(value.get("code"), str)
        and bool(value["code"].strip())
        and isinstance(value.get("message"), str)
        and bool(value["message"].strip())
    )


def _is_valid_snapshot(snapshot: Any, meeting_id: int) -> bool:
    if not isinstance(snapshot, dict):
        return False
    if (
        not _is_non_negative_int(snapshot.get("version"))
        or snapshot["version"] != 1
    ):
        return False
    if (
        not isinstance(snapshot.get("meeting_id"), str)
        or snapshot["meeting_id"] != str(meeting_id)
        or not isinstance(snapshot.get("native_meeting_id"), str)
        or not snapshot["native_meeting_id"].strip()
    ):
        return False
    if not all(
        _is_non_negative_int(snapshot.get(field))
        for field in (
            "updated_at_ms",
            "active_requests",
            "waiting_channels",
            "processed_windows",
            "superseded_windows",
        )
    ):
        return False
    if not all(
        _is_non_negative_number(snapshot.get(field))
        for field in (
            "active_audio_sec",
            "queued_audio_sec",
            "lag_sec",
        )
    ):
        return False
    if not all(
        _is_non_negative_number(snapshot.get(field), optional=True)
        for field in (
            "latest_captured_audio_end_ms",
            "latest_processed_audio_end_ms",
            "rtf_ema",
        )
    ):
        return False
    return _is_valid_last_error(snapshot.get("last_error"))


class AlloySttTelemetryReader:
    def __init__(self, redis: Any):
        self._redis = redis

    async def read_owned(self, meeting_ids: Iterable[int]) -> list[dict]:
        ids = [int(meeting_id) for meeting_id in meeting_ids]
        if not ids:
            return []
        raw_rows = await self._redis.mget(
            [alloy_stt_telemetry_key(meeting_id) for meeting_id in ids],
        )
        snapshots: list[dict] = []
        for meeting_id, raw in zip(ids, raw_rows):
            if not raw:
                continue
            try:
                snapshot = json.loads(raw)
            except (TypeError, ValueError, json.JSONDecodeError):
                continue
            if not _is_valid_snapshot(snapshot, meeting_id):
                continue
            snapshots.append(snapshot)
        return snapshots

File: core/meetings/modules/gmeet-pipeline/src/alloy-stt-telemetry.ts
/** ALLOY: local STT queue snapshot and request-lifecycle contract. */
export type AlloySttTelemetryError = {
  code: string;
  message: string;
};

export type AlloySttTelemetrySnapshotV1 = {
  version: 1;
  meeting_id: string;
  native_meeting_id: string;
  updated_at_ms: number;
  active_requests: number;
  active_audio_sec: number;
  waiting_channels: number;
  queued_audio_sec: number;
  latest_captured_audio_end_ms: number | null;
  latest_processed_audio_end_ms: number | null;
  lag_sec: number;
  rtf_ema: number | null;
  processed_windows: number;
  superseded_windows: number;
  last_error: AlloySttTelemetryError | null;
};

export interface AlloySttTelemetryTracker {
  captured(channelId: string, audioEndMs: number): void;
  queued(requestId: string, channelId: string, audioSec: number): void;
  superseded(requestId: string): void;
  started(requestId: string, channelId: string, audioSec: number): void;
  finished(requestId: string): void;
  completed(input: {
    requestId: string;
    audioSec: number;
    audioEndMs: number;
    executionDurationMs?: number;
  }): void;
  failed(requestId: string, error: AlloySttTelemetryError): void;
  recovered(): void;
  snapshot(): AlloySttTelemetrySnapshotV1;
}

/** ALLOY: request metadata only; PCM stays owned by the scheduler. */
type AlloySttRequestRecord = {
  channelId: string;
  audioSec: number;
};

const nonNegative = (value: number): number =>
  Number.isFinite(value) ? Math.max(0, value) : 0;

const sum = (values: Iterable<number>): number => {
  let total = 0;
  for (const value of values) total += value;
  return total;
};

export function createAlloySttTelemetryTracker(input: {
  meetingId: string;
  nativeMeetingId: string;
  now?: () => number;
}): AlloySttTelemetryTracker {
  const now = input.now ?? Date.now;
  // ALLOY: request identity keeps a limiter waiter and a newer scheduler
  // window on the same channel without either record overwriting the other.
  const pendingByRequest = new Map<string, AlloySttRequestRecord>();
  const activeByRequest = new Map<string, AlloySttRequestRecord>();

  let updatedAtMs = now();
  let latestCapturedAudioEndMs: number | null = null;
  let latestProcessedAudioEndMs: number | null = null;
  let rtfEma: number | null = null;
  let processedWindows = 0;
  let supersededWindows = 0;
  let lastError: AlloySttTelemetryError | null = null;

  const touch = () => {
    updatedAtMs = now();
  };

  const snapshot = (): AlloySttTelemetrySnapshotV1 => {
    const lagSec =
      latestCapturedAudioEndMs === null || latestProcessedAudioEndMs === null
        ? 0
        : Math.max(0, (latestCapturedAudioEndMs - latestProcessedAudioEndMs) / 1000);

    return {
      version: 1,
      meeting_id: input.meetingId,
      native_meeting_id: input.nativeMeetingId,
      updated_at_ms: updatedAtMs,
      active_requests: activeByRequest.size,
      active_audio_sec: sum([...activeByRequest.values()].map((record) => record.audioSec)),
      waiting_channels: new Set(
        [...pendingByRequest.values()].map((record) => record.channelId),
      ).size,
      queued_audio_sec: sum([...pendingByRequest.values()].map((record) => record.audioSec)),
      latest_captured_audio_end_ms: latestCapturedAudioEndMs,
      latest_processed_audio_end_ms: latestProcessedAudioEndMs,
      lag_sec: lagSec,
      rtf_ema: rtfEma,
      processed_windows: processedWindows,
      superseded_windows: supersededWindows,
      last_error: lastError,
    };
  };

  return {
    captured(_channelId, audioEndMs) {
      const normalized = nonNegative(audioEndMs);
      latestCapturedAudioEndMs =
        latestCapturedAudioEndMs === null
          ? normalized
          : Math.max(latestCapturedAudioEndMs, normalized);
      touch();
    },

    queued(requestId, channelId, audioSec) {
      pendingByRequest.set(requestId, {
        channelId,
        audioSec: nonNegative(audioSec),
      });
      touch();
    },

    superseded(requestId) {
      if (pendingByRequest.delete(requestId)) supersededWindows++;
      touch();
    },

    started(requestId, channelId, audioSec) {
      pendingByRequest.delete(requestId);
      activeByRequest.set(requestId, {
        channelId,
        audioSec: nonNegative(audioSec),
      });
      touch();
    },

    finished(requestId) {
      activeByRequest.delete(requestId);
      touch();
    },

    completed({ requestId, audioSec, audioEndMs, executionDurationMs }) {
      // ALLOY: custom/injected transcribers may ignore the optional observer.
      // Completion still drains their exact request without inventing execution time.
      pendingByRequest.delete(requestId);
      activeByRequest.delete(requestId);
      const normalizedEnd = nonNegative(audioEndMs);
      latestProcessedAudioEndMs =
        latestProcessedAudioEndMs === null
          ? normalizedEnd
          : Math.max(latestProcessedAudioEndMs, normalizedEnd);
      processedWindows++;

      if (
        Number.isFinite(audioSec)
        && audioSec > 0
        && executionDurationMs !== undefined
        && Number.isFinite(executionDurationMs)
      ) {
        const requestRtf = nonNegative(executionDurationMs) / 1000 / audioSec;
        rtfEma = rtfEma === null ? requestRtf : 0.2 * requestRtf + 0.8 * rtfEma;
      }
      touch();
    },

    failed(requestId, error) {
      // ALLOY: clear only the failed request; unrelated waiters remain visible.
      pendingByRequest.delete(requestId);
      activeByRequest.delete(requestId);
      lastError = error;
      touch();
    },

    recovered() {
      lastError = null;
      touch();
    },

    snapshot,
  };
}
